src/main/java/org/quiltmc/loader/impl/minecraft/McVersionLookup.java

```diff
@@ -65,63 +68,106 @@
 	private static final Pattern INDEV_PATTERN = Pattern.compile("(?:inf-|Inf?dev )(?:0\\.31 )?(\\d+(-\\d+)?)");
 	private static final String STRING_DESC = "Ljava/lang/String;";
 
-	public static McVersion getVersion(String version) {
-		return new McVersion(version, getRelease(version));
-	}
-
-	public static McVersion getVersion(Path gameJar) {
-		McVersion ret;
+	public static McVersion getVersion(Path gameJar, List<String> entrypointClasses, @Nullable String versionName) {
+		McVersion.Builder builder = new McVersion.Builder();
 
 		// check various known files for version information
+		if (versionName != null) {
+			builder.setNameAndRelease(versionName);
+		}
 
 		try (FileSystemUtil.FileSystemDelegate jarFs = FileSystemUtil.getJarFileSystem(gameJar, false)) {
 			FileSystem fs = jarFs.get();
+			// Determine class version
+			for (String entrypointClass : entrypointClasses) {
+				String fileString = entrypointClass.replace('.', '/') + ".class";
+				Path file = fs.getPath(fileString);
+
+				if (Files.isRegularFile(file)) {
+					try (DataInputStream is = new DataInputStream(Files.newInputStream(file))) {
+						if (is.readInt() != 0xCAFEBABE) {
+							continue;
+						}
+
+						is.readUnsignedShort();
+						builder.setClassVersion(is.readUnsignedShort());
```

Comment:
Really upstream, this could have just been a short lived class visitor...

Should we put a comment on this to rework it into a visitor?